reader.go

```diff
@@ -122,76 +122,56 @@ func (r *Reader) Config() ReaderConfig {
 // messages from it.
 func (r *Reader) Close() error {
 	r.mutex.Lock()
-	if !r.closed {
-		close(r.done)
-	}
+	closed := r.closed
 	r.closed = true
-	r.dirty = false
-	r.cancel()
 	r.mutex.Unlock()
+
+	r.cancel()
+	r.join.Wait()
```

Comment:
Yep, but I don't think it's really feasable. `sync.WaitGroup` doesn't expose the value of its internal counter, that would require a bunch of changes to the code to verify it... doesn't seem worth it.